rectools/models/bert4rec.py

```diff
@@ -0,0 +1,578 @@
+        return seqs
+
+    def forward(
+        self,
+        sessions: torch.Tensor,  # [batch_size, session_max_len]
+    ) -> torch.Tensor:
+        """TODO"""
+        item_embs = self.item_model.get_all_embeddings()  # [n_items + 2, n_factors]
+        session_embs = self.encode_sessions(sessions, item_embs)  # [batch_size, session_max_len, n_factors]
+        logits = session_embs @ item_embs.T  # [batch_size, session_max_len, n_items + 2]
+        return logits
+
+
+class SessionEncoderLightningModule(SessionEncoderLightningModuleBase):
+    """TODO"""
+
+    def on_train_start(self) -> None:
+        """TODO"""
+        self._truncated_normal_init()
+
+    def configure_optimizers(self) -> torch.optim.Adam:
+        """TODO"""
+        optimizer = torch.optim.Adam(self.torch_model.parameters(), lr=self.lr, betas=self.adam_betas)
+        return optimizer
+
+    def training_step(self, batch: torch.Tensor, batch_idx: int) -> torch.Tensor:
+        """TODO"""
+        x, y, w = batch
+        logits = self.forward(x)  # [batch_size, session_max_len, n_items + 2]
+        if self.loss == "softmax":
+            # We are using CrossEntropyLoss with a multi-dimensional case
+
+            # Logits must be passed in form of [batch_size, n_items + 2, session_max_len],
+            #  where n_items + 2 is number of classes
+
+            # Target label indexes must be passed in a form of [batch_size, session_max_len]
+            # (`0` index for "PAD" ix excluded from loss)
+
+            # Loss output will have a shape of [batch_size, session_max_len]
+            # and will have zeros for every `0` target label
+
+            loss = torch.nn.functional.cross_entropy(
+                logits.transpose(1, 2), y, ignore_index=0, reduction="none"
+            )  # [batch_size, session_max_len]
+            loss = loss * w
+            n = (loss > 0).to(loss.dtype)
+            loss = torch.sum(loss) / torch.sum(n)
+            return loss
+        raise ValueError(f"loss {loss} is not supported")
+
+    def _truncated_normal_init(self) -> None:
+        """TODO"""
+        for _, param in self.torch_model.named_parameters():
+            try:
+                torch.nn.init.trunc_normal_(param.data)
+            except ValueError:
+                pass
+
+
+class BERT4RecModel(ModelBase):
+    """TODO"""
+
+    def __init__(  # pylint: disable=too-many-arguments, too-many-locals
+        self,
+        n_blocks: int = 1,
+        n_heads: int = 1,
+        n_factors: int = 128,
+        use_pos_emb: bool = True,
+        dropout_rate: float = 0.2,
+        epochs: int = 3,
+        verbose: int = 0,
+        deterministic: bool = False,
+        cpu_n_threads: int = 0,
+        session_max_len: int = 32,
+        batch_size: int = 128,
+        loss: str = "softmax",
+        lr: float = 0.01,
+        dataloader_num_workers: int = 0,
+        train_min_user_interaction: int = 2,
+        mask_prob: float = 0.15,
```

Comment:
As far as I can see the only difference in this model from SASRec is receiving mask_prob and passing it to data_preparator_type.
What you need to do:
1. Create `TransformerModelBase` class with all of the common methods already implemented. Look here https://github.com/MobileTeleSystems/RecTools/blob/main/rectools/models/base.py for example. Carefully select which of the arguments must be passed to __init__. `mask_prob` shouldn't be there at all. `transformer_layers_type` and `data_preparator_type` should not have default values. `self.data_preparator` should not be initialized but should be declared like `self.data_preparator: SessionEncoderDataPreparatorBase`. 
2. In SASRecModel __init__ create default values for `transformer_layers_type` and `data_preparator_type`. Initialize self.data_preparator
3. In BERT4RecModel __init__ create default values for `transformer_layers_type` and `data_preparator_type`. Initialize self.data_preparator

(This is a draft for now. I will update it soon. Don't implement for now)

Reply:
fixed